File: edit-distance/edit_distance.clj
(defn edit-distance [str1 str2]
  (let [m (count str1)
        n (count str2)
        dp (vec (for [i (range (inc m))]
                 (vec (repeat (inc n) 0))))]
    
    (let [dp (reduce (fn [dp i] (assoc-in dp [i 0] i)) dp (range (inc m)))
          dp (reduce (fn [dp j] (assoc-in dp [0 j] j)) dp (range (inc n)))]
      
      (loop [dp dp i 1]
        (if (> i m)
          (get-in dp [m n])
          (recur
            (loop [dp dp j 1]
              (if (> j n)
                dp
                (recur
                  (if (= (nth str1 (dec i)) (nth str2 (dec j)))
                    (assoc-in dp [i j] (get-in dp [(dec i) (dec j)]))
                    (assoc-in dp [i j] (inc (min (get-in dp [(dec i) j])
                                                (get-in dp [i (dec j)])
                                                (get-in dp [(dec i) (dec j)])))))
                  (inc j))))
            (inc i)))))))

(println (edit-distance "kitten" "sitting"))

File: edit-distance/EditDistance.kt
fun editDistance(str1: String, str2: String): Int {
    val m = str1.length
    val n = str2.length
    val dp = Array(m + 1) { IntArray(n + 1) }
    
    for (i in 0..m) dp[i][0] = i
    for (j in 0..n) dp[0][j] = j
    
    for (i in 1..m) {
        for (j in 1..n) {
            dp[i][j] = if (str1[i-1] == str2[j-1]) {
                dp[i-1][j-1]
            } else {
                1 + minOf(dp[i-1][j], dp[i][j-1], dp[i-1][j-1])
            }
        }
    }
    
    return dp[m][n]
}

fun main() {
    println(editDistance("kitten", "sitting"))
}
